path_dynamic_roles.go

```diff
@@ -0,0 +1,347 @@
+			HelpDescription: "List all the dynamic roles being managed by Vault.",
+		},
+		// GET credentials
+		{
+			Pattern: path.Join(dynamicCredPath, framework.MatchAllRegex("name")),
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeLowerCaseString,
+					Description: "Name of the dynamic role.",
+				},
+			},
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ReadOperation: &framework.PathOperation{
+					Callback: b.pathDynamicCredsRead,
+				},
+			},
+			HelpSynopsis: "Request LDAP credentials for a dynamic role. These credentials are " +
+				"created within OpenLDAP when querying this endpoint.",
+			HelpDescription: "This path requests new LDAP credentials for a certain dynamic role. " +
+				"The credentials are created within OpenLDAP based on the creation_ldif specified " +
+				"within the dynamic role configuration.",
+		},
+	}
+}
+
+func dynamicSecretCreds(b *backend) *framework.Secret {
+	return &framework.Secret{
+		Type: secretCredsType,
+		Fields: map[string]*framework.FieldSchema{
+			"username": {
+				Type:        framework.TypeString,
+				Description: "Username of the generated account",
+			},
+			"password": {
+				Type:        framework.TypeString,
+				Description: "Password to access the generated account",
+			},
+			"distinguished_names": {
+				Type: framework.TypeStringSlice,
+				Description: "List of the distinguished names (DN) created. Each name in this list corresponds to" +
+					"each action taken within the creation_ldif statements. This does not de-duplicate entries, " +
+					"so this will have one entry for each LDIF statement within creation_ldif.",
+			},
+		},
+
+		Renew:  b.secretCredsRenew(),
+		Revoke: b.secretCredsRevoke(),
+	}
+}
+
+func (b *backend) pathDynamicRoleCreateUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	rawData := data.Raw
+	err := convertToDuration(rawData, "default_ttl", "max_ttl")
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert TTLs to duration: %w", err)
+	}
+
+	dRole := &dynamicRole{}
+	err = mapstructure.WeakDecode(rawData, dRole)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode request: %w", err)
+	}
+
+	dRole.CreationLDIF = decodeBase64(dRole.CreationLDIF)
+	dRole.RollbackLDIF = decodeBase64(dRole.RollbackLDIF)
+	dRole.DeletionLDIF = decodeBase64(dRole.DeletionLDIF)
+
+	err = validateDynamicRole(dRole)
+	if err != nil {
+		return nil, err
+	}
+
+	err = storeDynamicRole(ctx, req.Storage, dRole)
+	if err != nil {
+		return nil, fmt.Errorf("failed to save dynamic role: %w", err)
+	}
+
+	return nil, nil
+}
+
+func validateDynamicRole(dRole *dynamicRole) error {
+	merr := new(multierror.Error)
+
+	merr = multierror.Append(merr,
+		firstError(
+			errIfTrue(dRole.CreationLDIF == "", "missing creation_ldif"),
+			errWrap(assertValidLDIFTemplate(dRole.CreationLDIF), "invalid creation_ldif: %w"),
+		),
+	)
```

Comment:
This feels like a bit too much error-handling magic without having to dive into what each of these helpers are doing.

Can we simplify the logic here to something like this?
```
var creationErr error
if dRole.CreationLDIF == "" {
  creationErr = errors.New("missing creation_ldif")
}
if creationErr == nil {
    if err := assertValidLDIFTemplate(dRole.CreationLDIF); err != nil {
      creationErr = fmt.Errorf("invalid creation_ldif: %w", err)
    }
}
merr = multierror.Append(merr, creationErr) 
```

Reply:
I disagree that this will be more readable and more maintainable than other options. This code would need to be copy-pasta'd several times with small changes between them. This will result in future maintainability problems because copy-pasta of large blocks with minor changes is prone to human error.

I've changed this to a fail fast model so we can get this PR merged.

Reply:
I'm good with this approach since that's what we default for these types of checks and it keeps the logic clean.